c++/src/kj/compat/http.c++

```diff
@@ -4084,95 +4129,26 @@ public:
-              auto parameters = _::splitPartsWithCondition(finalOffer, ';');
-              auto extName = parameters[0].asArray();
-              // First, let's make sure the extension doesn't have invalid trailing characters.
-              ArrayPtr<const char> extNamePtr = extName.slice(permittedExtension.size(),
-                                                              extName.size());
-              _::stripLeadingSpace(extNamePtr);
-
-              if (extNamePtr.size() > 0) {
-                // Invalid characters after extension name, we'll decline the agreement.
-                auto msg = kj::str(
-                    failed,
-                    "response included a Sec-WebSocket-Extensions value that was "
-                    "not permessage-deflate.");
-                return settings.errorHandler.orDefault(*this).handleWebSocketProtocolError({
-                  502, "Bad Gateway", msg, nullptr
-                });
-              }
-              if (parameters.size() > 1) {
-                auto keyMaybeValuePairs = _::toKeysAndVals(parameters.slice(1, parameters.size()));
-                auto maybeUnverified = _::populateUnverifiedConfig(keyMaybeValuePairs);
-                // If we cannot construct an `UnverifiedConfig`, we'll return an error below.
-                KJ_IF_MAYBE(unverified, maybeUnverified) {
-                  // Agreed Parameters structure is generally correct, now we need to check that the
-                  // values are valid.
-                  compressionParameters = _::validateCompressionConfig(kj::mv(*unverified), true);
+              KJ_SWITCH_ONEOF(_::tryParseExtensionAgreement(offeredExtensions, *agreedParameters)) {
+                KJ_CASE_ONEOF(exception, kj::Exception) {
+                  return settings.errorHandler.orDefault(*this).handleWebSocketProtocolError({
+                    502, "Bad Gateway", exception.getDescription(), nullptr});
+                }
+                KJ_CASE_ONEOF(config, CompressionParameters) {
+                  compressionParameters = kj::mv(config);
                 }
-              } else {
-                // Plain `permessage-deflate`.
-                compressionParameters = CompressionParameters{};
-              }
-
-              if (compressionParameters == nullptr) {
-                // There was a problem parsing the server's `Sec-WebSocket-Extensions` response.
-                auto msg = kj::str(
-                    failed,
-                    "the Sec-WebSocket-Extensions header in the Response included "
-                    "an invalid value.");
-                return settings.errorHandler.orDefault(*this).handleWebSocketProtocolError({
-                  502, "Bad Gateway", msg, nullptr
-                });
               }
-              auto& agreement = KJ_REQUIRE_NONNULL(compressionParameters);
+              auto& agreement = KJ_ASSERT_NONNULL(compressionParameters);
```

Comment:
Seems like this line is causing some trouble: `C2065: 'compressionParameters': undeclared identifier`.

Just noticed the "_one `KJ_SWITCH_ONEOF()` per block_" in C++14, will fix this in the next fixup.

Reply:
This still seems weird, I don't see why nested switch statements would be a problem, but whatever, this seems fine.